File: library/src/main/java/su/levenetc/android/textsurface/animations/ScaleSurface.java
package su.levenetc.android.textsurface.animations;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.animation.PropertyValuesHolder;
import android.animation.ValueAnimator;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.animation.FastOutSlowInInterpolator;

import su.levenetc.android.textsurface.contants.Pivot;
import su.levenetc.android.textsurface.SurfaceCamera;
import su.levenetc.android.textsurface.Text;
import su.levenetc.android.textsurface.TextSurface;
import su.levenetc.android.textsurface.utils.Utils;
import su.levenetc.android.textsurface.interfaces.ICameraAnimation;
import su.levenetc.android.textsurface.interfaces.IEndListener;

/**
 * Created by Eugene Levenetc.
 */
public class ScaleSurface implements ICameraAnimation, ValueAnimator.AnimatorUpdateListener {

	private SurfaceCamera camera;
	private TextSurface textSurface;
	private int duration;
	private final Text textPivot;
	private int fit = -1;
	private int pivot;
	private float toScale;
	private ObjectAnimator animator;

	public ScaleSurface(int duration, Text textPivot, int pivot, float toScale) {
		this.duration = duration;
		this.textPivot = textPivot;
		this.pivot = pivot;
		this.toScale = toScale;
	}

	public ScaleSurface(int duration, Text textPivot, int fit) {
		this.duration = duration;
		this.textPivot = textPivot;
		this.fit = fit;
	}

	@Override public void setCamera(SurfaceCamera camera) {
		this.camera = camera;
	}

	@Override public void onStart() {

	}

	@Override public void start(@Nullable IEndListener listener) {

		float pivotX;
		float pivotY;

		if (fit == -1) {
			pivotX = textPivot.getPosition().getRelativeX(pivot, textPivot, true);
			pivotY = textPivot.getPosition().getRelativeY(pivot, textPivot, true);
		} else {
			int surfaceWidth = textSurface.getWidth();
			float textWidth = textPivot.getWidth();
			toScale = surfaceWidth / textWidth;
			pivotX = textPivot.getPosition().getRelativeX(Pivot.CENTER, textPivot, true);
			pivotY = textPivot.getPosition().getRelativeY(Pivot.CENTER, textPivot, true);
		}

		PropertyValuesHolder scaleHolder = PropertyValuesHolder.ofFloat("scale", camera.getScale(), toScale);
		PropertyValuesHolder pivotXHolder = PropertyValuesHolder.ofFloat("scalePivotX", camera.getScalePivotX(), pivotX);
		PropertyValuesHolder pivotYHolder = PropertyValuesHolder.ofFloat("scalePivotY", camera.getScalePivotY(), pivotY);

		animator = ObjectAnimator.ofPropertyValuesHolder(camera, scaleHolder, pivotXHolder, pivotYHolder);
		animator.setInterpolator(new FastOutSlowInInterpolator());
		animator.setDuration(duration);
		animator.addUpdateListener(this);
		Utils.addEndListener(this, animator, listener);
		animator.start();
	}

	@Override public void setTextSurface(@NonNull TextSurface textSurface) {
		this.textSurface = textSurface;
	}

	@Override public long getDuration() {
		return duration;
	}

	@Override public void cancel() {
		if (animator != null && animator.isRunning()) {
			animator.cancel();
		}
	}

	@Override public void onAnimationUpdate(ValueAnimator animation) {
		textSurface.invalidate();
	}

	@Override public String toString() {
		return "ScaleSurface{" +
				"textPivot=" + ((textPivot == null) ? "null" : textPivot.toString()) +
				'}';
	}
}

File: library/src/main/java/su/levenetc/android/textsurface/animations/ScaleSurface.kt
package su.levenetc.android.textsurface.animations

import android.animation.ObjectAnimator
import android.animation.PropertyValuesHolder
import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.annotation.SuppressLint
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import su.levenetc.android.textsurface.SurfaceCamera
import su.levenetc.android.textsurface.Text
import su.levenetc.android.textsurface.TextSurface
import su.levenetc.android.textsurface.contants.Pivot
import su.levenetc.android.textsurface.interfaces.ICameraAnimation
import su.levenetc.android.textsurface.interfaces.IEndListener
import su.levenetc.android.textsurface.utils.Utils.addEndListener

/**
 * Created by Eugene Levenetc.
 */
class ScaleSurface : ICameraAnimation, AnimatorUpdateListener {
    private var camera: SurfaceCamera? = null
    override lateinit var textSurface: TextSurface
    override var duration: Long
    private val textPivot: Text?
    private var fit = -1
    private var pivot = 0
    private var toScale = 0f
    private var animator: ObjectAnimator? = null

    constructor(duration: Long, textPivot: Text?, pivot: Int, toScale: Float) {
        this.duration = duration
        this.textPivot = textPivot
        this.pivot = pivot
        this.toScale = toScale
    }

    constructor(duration: Long, textPivot: Text?, fit: Int) {
        this.duration = duration
        this.textPivot = textPivot
        this.fit = fit
    }

    override fun setCamera(camera: SurfaceCamera?) {
        this.camera = camera
    }

    override fun onStart() {}

    @SuppressLint("ObjectAnimatorBinding")
    override fun start(listener: IEndListener) {
        val pivotX: Float
        val pivotY: Float
        if (fit == -1) {
            pivotX = textPivot!!.position.getRelativeX(pivot, textPivot, true)
            pivotY = textPivot.position.getRelativeY(pivot, textPivot, true)
        } else {
            val surfaceWidth = textSurface!!.width
            val textWidth = textPivot!!.width
            toScale = surfaceWidth / textWidth
            pivotX = textPivot.position.getRelativeX(Pivot.CENTER, textPivot, true)
            pivotY = textPivot.position.getRelativeY(Pivot.CENTER, textPivot, true)
        }
        val scaleHolder = PropertyValuesHolder.ofFloat("scale", camera!!.getScale(), toScale)
        val pivotXHolder = PropertyValuesHolder.ofFloat("scalePivotX", camera!!.scalePivotX, pivotX)
        val pivotYHolder = PropertyValuesHolder.ofFloat("scalePivotY", camera!!.scalePivotY, pivotY)
        animator = ObjectAnimator.ofPropertyValuesHolder(camera, scaleHolder, pivotXHolder, pivotYHolder)
        animator!!.interpolator = FastOutSlowInInterpolator()
        animator!!.duration = duration.toLong()
        animator!!.addUpdateListener(this)
        addEndListener(this, animator!!, listener)
        animator!!.start()
    }

    override fun cancel() {
        if (animator != null && animator!!.isRunning) {
            animator!!.cancel()
        }
    }

    override fun onAnimationUpdate(animation: ValueAnimator) {
        textSurface!!.invalidate()
    }

    override fun toString(): String {
        return "ScaleSurface{" +
                "textPivot=" + (textPivot?.toString() ?: "null") +
                '}'
    }
}